### PR TITLE
TLS configuration for Slack service for Mattermost compatibility

### PR DESCRIPTION
**Pull request by seuf**
Hi,
Mattermost is compatible with slack messages json format so I've patched the slack services to add TLS configuration parameters.
Now alerts can be sent to self hosted Mattermost with custom certificates.

###### Required for all non-trivial PRs
- [X] Rebased/mergable
- [x] Tests pass
- [x] CHANGELOG.md updated
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -475,7 +475,10 @@ func (s *Server) appendSensuService() {
 func (s *Server) appendSlackService() {
 	c := s.config.Slack
 	l := s.LogService.NewLogger("[slack] ", log.LstdFlags)
-	srv := slack.NewService(c, l)
+	srv, err := slack.NewService(c, l)
+	if err != nil {
+		s.Logger.Println("W! Slack init :", err)
```
Comment (nathanielc) [inline server/server.go]: Go ahead and return the error here and check it in NewServer and return it from there. That will cause the server to print the error message and shutdown, which is the behavior we want.

File: services/slack/service.go [hunk 2]
```diff
@@ -17,14 +19,21 @@ import (
...
 	logger      *log.Logger
+	client      *http.Client
 }
 
-func NewService(c Config, l *log.Logger) *Service {
+func NewService(c Config, l *log.Logger) (*Service, error) {
+	tlsConfig, err := getTLSConfig(c.SSLCA, c.SSLCert, c.SSLKey, c.InsecureSkipVerify)
+	if err != nil {
+		return nil, err
+	}
```
Comment (nathanielc) [inline services/slack/service.go]: Can you add a log warn here if the config specifies using InsecureSkipVerify?

File: services/slack/service.go [hunk 3]
```diff
@@ -17,14 +19,21 @@ import (
...
 	s := &Service{
 		logger: l,
+		client: &http.Client{Transport: &http.Transport{TLSClientConfig: tlsConfig}},
```
Comment (nathanielc) [inline services/slack/service.go]: The configuration can be changed at runtime via the API, there should be an Update method on the service which is in charge of applying new config changes. Can you update that method to allow for creating a new client when the TLS config changes. See the Alerta service for an example.

**Comment by seuf** [comment]
I've updated the pr to add all your requested changes. Is it good to you ?

File: services/slack/service.go [hunk 4]
```diff
@@ -99,7 +119,8 @@ func (s *Service) Alert(channel, message, username, iconEmoji string, level aler
 	if err != nil {
 		return err
 	}
-	resp, err := http.Post(url, "application/json", post)
+	s.logger.Println("D! Posting event to Slack", post)
+	resp, err := s.client.Post(url, "application/json", post)
```
Comment (nathanielc) [inline services/slack/service.go]: This line races with the lines in Update and New that assign to the client. Either add a mutex to the Sensu service or use the sync.Atomic type to synchronizes access to the client.

Other than that this looks good! Thanks

Comment (seuf) [inline services/slack/service.go]: Hum. I'm not a Go dev. This looks like a bit tricky for me.. O_ô 
I will try to do this next tuesday when i go back to work.

Comment (nathanielc) [inline services/slack/service.go]: @seuf No worries, I can help you out too. The Alerta service does something similar if that helps.

Comment (seuf) [inline services/slack/service.go]: OK, I have updated the PR to use atomic value for the http client like Alerta service.

File: services/slack/service.go [hunk 5]
```diff
@@ -99,7 +127,9 @@ func (s *Service) Alert(channel, message, username, iconEmoji string, level aler
 	if err != nil {
 		return err
 	}
-	resp, err := http.Post(url, "application/json", post)
+	s.logger.Println("D! Posting event to Slack", post)
```
Comment (nathanielc) [inline services/slack/service.go]: Can you remove this debug statement?

**Comment by seuf** [comment]
done

**Comment by nathanielc** [comment]
@seuf Thanks! :tada:

**Pull request merged**